samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -88,21 +88,26 @@ class KafkaCheckpointManager(
     val msgBytes = serde.toBytes(checkpoint)
     retryBackoff.run(
       loop => {
-        if (producer == null) {
-          producer = connectProducer()
+        if (systemProducer == null) {
+          systemProducer = getSystemProducer()
+          systemProducer.register(taskName.getTaskName)
+          systemProducer.start
         }
 
-        producer.send(new ProducerRecord(checkpointTopic, 0, keyBytes, msgBytes)).get()
+        val systemStream = new SystemStream(systemName, checkpointTopic)
+        val envelope: OutgoingMessageEnvelope = new OutgoingMessageEnvelope(systemStream, keyBytes, msgBytes)
+        systemProducer.send(taskName.getTaskName, envelope)
+        systemProducer.flush(taskName.getTaskName) // make sure it is written
         loop.done
       },
 
       (exception, loop) => {
         warn("Failed to write %s partition entry %s: %s. Retrying." format(CHECKPOINT_LOG4J_ENTRY, key, exception))
         debug("Exception detail:", exception)
-        if (producer != null) {
-          producer.close
+        if (systemProducer != null) {
+          systemProducer.stop
```

Comment:
iiuc, multiple threads can't call write checkpoint on the checkpoint manager right? So, there's no synchronization needed on the producer? (even with async commit enabled)
cc: @prateekm to confirm

Reply:
Good catch @vjagadish1989, this is a bug. We can have concurrent commit for concurrently executing TaskInstances, even without async commit enabled. We need to make this producer recycling thread safe.

If I'm reading this correctly, we retry forever until we succeed? Since there aren't that many errors that can be fixed by recreating the producer, we probably don't need to recycle it (should verify).

Reply:
hmm, since it is existing code, I guess this problem existed before. 
If we keep recycling, then the only way would be to synchronize the whole thing (all accesses to the producer),
but if we remove the recycling then we only need to sync the initialization (to avoid double creation).
Hopefully the new producer is resilient enough and doesn't need to be recycled.